File: python/linked_list/160. Intersection of Two Linked Lists.py
'''
Two pointer
ptr1: headA => headB
ptr2: headB => headA
Time: O(m + n)
Space: O(1)
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        ptr1, ptr2 = headA, headB

        while ptr1 != ptr2:
            ptr1 = ptr1.next if ptr1 else headB
            ptr2 = ptr2.next if ptr2 else headA

        return ptr1

File: javascript/linked_list/160. Intersection of Two Linked Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    let ptr1 = headA, ptr2 = headB;

    while (ptr1 != ptr2) {
        ptr1 ? ptr1 = ptr1.next : ptr1 = headB;
        ptr2 ? ptr2 = ptr2.next : ptr2 = headA;
    }
    return ptr1;
};